framework/src/utils/DelimitedFileReader.C

```diff
@@ -0,0 +1,184 @@
+  raw.resize(size_raw);
+  _communicator->broadcast(raw);
+
+  // Update the data
+  _data.resize(n_cols);
+  const std::size_t n_rows = raw.size() / n_cols;
+  for (std::size_t j = 0; j < n_cols; ++j)
+  {
+    _data[j].resize(n_rows);
+    for (std::size_t i = 0; i < n_rows; ++i)
+      _data[j][i] = raw[i * n_cols + j];
+  }
+}
+
+const std::vector<std::string> &
+DelimitedFileReader::getColumnNames()
+{
+  return _column_names;
+}
+
+const std::vector<std::vector<double>> &
+DelimitedFileReader::getColumnData()
+{
+  return _data;
+}
+
+void
+DelimitedFileReader::initializeColumns(std::ifstream & stream_data)
+{
+  // Storage for line content
+  std::string line;
+
+  // Read the header (this is the default)
+  if (_header)
+  {
+    std::getline(stream_data, line);
+    MooseUtils::tokenize(line, _column_names, 1, _delimiter);
+    for (std::string & str : _column_names)
+      str = MooseUtils::trim(str);
+  }
+
+  // Generate the header
+  else
+  {
+    // Read the first line of data (to set expected column size) and return back to original pos
+    std::streampos pos = stream_data.tellg();
+    std::getline(stream_data, line);
+    MooseUtils::tokenize(line, _column_names, 1, _delimiter);
+    stream_data.seekg(pos);
+
+    // Create names
+    std::size_t n = _column_names.size();
+    int padding = std::log(n);
+    for (std::size_t i = 0; i < n; ++i)
+    {
+      std::stringstream ss;
+      ss << "column_" << std::setw(padding) << std::setfill('0') << i;
+      _column_names[i] = ss.str();
+    }
+  }
+}
+
+void
+DelimitedFileReader::readData(std::ifstream & stream_data, std::vector<double> & output)
+{
+  // Local storage for the data being read
+  std::string line;
+  std::vector<double> row;
+
+  // Keep track of the row number for error reporting
+  unsigned int count = _header ? 1 : 0;
+
+  // The number of columns expected based on the first row of the data
+  const size_t n_cols = _column_names.size();
+
+  // Read the lines
+  while (std::getline(stream_data, line))
+  {
+    // Increment row counter and clear any tokenized data
+    count++;
+    row.clear();
+
+    // Separate the row and error if it fails
+    bool status = MooseUtils::tokenizeAndConvert<double>(line, row, _delimiter);
+    if (!status)
+      mooseError("Failed to convert a delimited data into double when reading row ",
+                 count,
+                 " in file ",
+                 _filename,
+                 ".");
+
+    // Check number of columns
+    if (row.size() != n_cols)
+      mooseError("The number of columns read (",
+                 row.size(),
+                 ") does not match the number of columns expected (",
+                 n_cols,
+                 ") based on the first row of the file when reading row ",
+                 count,
+                 " in file ",
+                 _filename,
+                 ".");
+
+    // Append data
+    output.insert(output.end(), row.begin(), row.end());
```

Comment:
You can't know the size because we don't know the number of rows. I could go through the file and get the number of lines, then go back and read the data. Not sure if this is better.